### DISCUSSION TITLE
Changing of the hyperreduction methods

### DISCUSSION THREAD

**Discussion by giovastabile** [ITHACA-FV/ITHACA-FV#510] [Ideas]
First message

**Comment by vressegu** [comment]
The core of the **DEIM algorithm could be implemented in a standard method of the DEIM class, instead of being implemented in the constructor of the DEIM class.**

Then, the DEIM algorithm would be call either (i) from the constructor or (ii) from a dedicated method.
The solution (i) would not how to use the DEIM class (no change in the tutorials (8 and 9), and no changes in homemade codes of ITHACA-FV users). But the solution (ii) may be more relevant in term of structure of the code.

This implementation change will facilitate the implementation of classes inheriting from the DEIM class.

**Comment by vressegu** [comment]
The DEIM constructor / DEIM algorithm could use already computed modes, instead of snapshots.

**Reply by vressegu** [reply]
I have some code to fix this, but first I need separated DEIM algorithm and DEIM class constructor.

**Comment by vressegu** [comment]
The DEIM algorithm (magic points choice) could better deal with volVectorFields.

**Reply by vressegu** [reply]
We also have some code to fix this (and associated theoretical bounds of interpolation errors),
but first we need separated DEIM algorithm and DEIM class constructor.

**Reply by vressegu** [reply]
The DEIM algorithm (magic points choice) could better deal with vectors of more than 3 components (e.g., (velocity,pressure)). One solution could be to use volTensorFields.

**Comment by vressegu** [comment]
The DEIM algorithm (magic points choice) may better deal with non-uniform mesh (best interpolation in term of L2).

**Reply by vressegu** [reply]
We also have some code to fix this (and associated theoretical bounds of interpolation errors),
but first we need separated DEIM algorithm and DEIM class constructor.

**Reply by fraromor** [reply]
Can we deal with this outside the HR class, when computing the basis for HR ?

**Reply by vressegu** [reply]
I think so, even though our code was not written that way.
I think the 2 requirements would be:
(i) Computing a basis optimal in term of L2 (typically considering volumes when computing the mass matrix to be diagonalized).
(ii) Providing a weighted basis to the HR algorithm (weighted by sqrt(dV_k) where dV_k is the volume of the k-th mesh element). As such the iterative choice of magic point will be optimal in term of L2 norm (and not Frobenius norm).

**Comment by fraromor** [comment]
Base class to implement both interpolation-based (I-HR) and numerical quadrature-based (Q-HR) hyper-reduction methods.
Main inputs: a function that returns hyper-reduction basis, initial magic points, mesh
Main outputs: numerical quadrature weights (=1 for interpolation-based methods), magic points, submesh

Separation of hyper-reduction basis computation and magic points (+ numerical quadrature weights) evaluation inside the hyper-reduction class.

Possibility to pass directly all the basis of hyper-reduction for small problems (e.g. from SVD or rSVD of residual ROM snapshots) or pass them sequentially in an efficient way (e.g. basis from sequential SVD algorithms applied to residual ROM snapshots). This feature can be implemented later.

Implement for both interpolation-based and numerical quadrature-based methods the possibility to perform HR on a single field or on lhs matrix + rhs field: examples of I-HR on a single field include (DEIM, gappy-POD) and on lhs + rhs include (GNAT); examples of Q-HR on a single field include (EQP, ECP) and on lhs + rhs (ECSW).

For I-HR implement directly over-sampling since it is more general w.r.t. forcing the number of magic points to be equal to the number of HR basis.

Of course, everything includes the case of stacked physical fields (e.g. pressure+velocity): proper normalization w.r.t. physical quantities and non-uniform mesh sizes should be dealt with when computing the basis of HR.

Implement greedy methods and S-OPT for I-HR.

Implement l1-norm optimization (EQP) and l2-norm optimization (ECP) for Q-HR. Since in some versions of EQP and ECP the computational cost is high, offer the possibility to choose between a fast but less accurate implementation and a slow but accurate one.

Be clear in the separation of Offline and Online stages.

References:
- DEIM, Chaturantabut, Saifon, and Danny C. Sorensen. "Nonlinear model reduction via discrete empirical interpolation." SIAM Journal on Scientific Computing 32.5 (2010): 2737-2764.
- GNAT, Carlberg, Kevin, et al. "The GNAT method for nonlinear model reduction: effective implementation and application to computational fluid dynamics and turbulent flows." Journal of Computational Physics 242 (2013): 623-647.
- ECP, Hernandez, Joaquin Alberto, Manuel Alejandro Caicedo, and Alex Ferrer. "Dimensional hyper-reduction of nonlinear finite element models via empirical cubature." Computer methods in applied mechanics and engineering 313 (2017): 687-722.
- ECSW, Farhat, Charbel, Todd Chapman, and Philip Avery. "Structure‐preserving, stability, and accuracy properties of the energy‐conserving sampling and weighting method for the hyper reduction of nonlinear finite element dynamic models." International journal for numerical methods in engineering 102.5 (2015): 1077-1110.
- EQP, Yano, Masayuki. "Discontinuous Galerkin reduced basis empirical quadrature procedure for model reduction of parametrized nonlinear conservation laws." Advances in Computational Mathematics 45.5 (2019): 2287-2320.

**Comment by MatthieuPetitS** [comment]
#### Hyper Reduction branch modifications
1. **Modifications of the tutorial 08DEIM to use the hyper-reduction class.** 
    1. The hyper-reduction class has some C++17 functionalities so it is necessary to adapt the compilation in the Make/options (add `-std=c++17`)
    2.  I added the MatrixOnline member to the hyper-reduction class and calculated it inside the GappyDEIM method. It can also be done in the main (which will allow for example the use of other methods such as the ECP).
    3. To keep the same folder hierarchy as before, I had to overload the GappyDEIM method to be able to change the saving folder. This is a main concern that can be useful to users, to choose the folder in which they want to save data. In the actual version, the folder ITHACAoutput is not exactly the same as before but it can be done with the overloaded method.
    4. Changes can be seen in Valentin's Fork in the devMatthieu branch. Tutorial 08DEIM : `HR08.DEIM.C`, Class `hyperReduction.H` and `hyperReduction.templates.H`) : https://github.com/vressegu/ITHACA-FV/tree/devMatthieu
2. **Ideas / Problems:**
    1. It might be useful for users to choose in which folder they want to save data. In the hyper-reduction methods (GappyDEIM, ECP...) and the modes calculation (SVD, ...)
    2. Creating .C and .H files for the hyper-reduction class is difficult due to the variadic template class. Is it necessary to do the same process as for the other classes to create a .so library ? If you have a solution to do it I'm interested.
    3. Write a new constructor so we don't have to pass the snapshots or the modes ?
    4. Let the user choose which inversion method he wants to use (inside the IthacaDict?) ?

**Reply by vressegu** [reply]
Matthieu found a workaround for issue 2.iii.

2.i and 2.iv are possible improvements that can be discussed after.

We do not have a solution for the issue 2.ii.
Do you have one? Or, are you ok for having a class HR without associated .so library?

More generally, 
do we need additional modifications before the pull request?
We can do a visio for presenting the code modifications if needed.

**Reply by giovastabile** [reply]
Hi Valentin, thanks. I think we can schedule a meeting to see what is missing together. Let me know when it works for you.

**Comment by vressegu** [comment]
Hi Giovanni, 

Are you available Monday 8th or Tuesday 9th of April? 

Valentin 



De: "Giovanni Stabile" ***@***.***> 
À: "ITHACA-FV/ITHACA-FV" ***@***.***> 
Cc: "RESSEGUIER Valentin" ***@***.***>, "Comment" ***@***.***> 
Envoyé: Lundi 25 Mars 2024 09:14:28 
Objet: Re: [ITHACA-FV/ITHACA-FV] Changing of the hyperreduction methods (Discussion #510) 





Hi Valentin, thanks. I think we can schedule a meeting to see what is missing together. Let me know when it works for you.

**Reply by giovastabile** [reply]
Hi @vressegu both options work for me! What about 10 on Monday morning?
